File: src/mobility/MobilityManager.cpp
/**
 * @file MobilityManager.cpp
 * @brief Implementation of MobilityManager.
 */

#include "mobility/MobilityManager.hpp"

#include <xbt/log.h>

#include <filesystem>
#include <fstream>
#include <iomanip>
#include <set>
#include <sstream>
#include <stdexcept>

XBT_LOG_NEW_DEFAULT_CATEGORY(mobility_manager, "ENIGMA Mobility Manager");

namespace fs = std::filesystem;

namespace enigma::mobility {

// ------------------------------------------------------------------ //
// helpers
// ------------------------------------------------------------------ //

static std::string quote(const std::string& s) {
    std::string out = "\"";
    for (char c : s) {
        if (c == '"') out += "\\\"";
        else          out += c;
    }
    return out + "\"";
}

// ------------------------------------------------------------------ //
// Constructors
// ------------------------------------------------------------------ //

std::string MobilityManager::detect_mobility_dir(simgrid::s4u::Engine& engine) const {
    // Walk all zones and look for the "mobility_dir" property
    for (auto* zone : engine.get_all_netzones()) {
        const char* val = zone->get_property("mobility_dir");
        if (val && val[0] != '\0') {
            XBT_INFO("Found mobility_dir property in zone '%s': %s",
                     zone->get_cname(), val);
            return std::string(val);
        }
    }
    // Also check host-level properties
    for (auto* host : engine.get_all_hosts()) {
        const char* val = host->get_property("mobility_dir");
        if (val && val[0] != '\0') {
            XBT_INFO("Found mobility_dir property on host '%s': %s",
                     host->get_cname(), val);
            return std::string(val);
        }
    }
    return {};
}

MobilityManager::MobilityManager(simgrid::s4u::Engine& engine) {
    std::string dir = detect_mobility_dir(engine);
    if (dir.empty()) {
        XBT_INFO("No mobility_dir property found – mobility module inactive");
        return;
    }
    coords_dir_ = dir;
    load_directory(dir);
}

MobilityManager::MobilityManager(simgrid::s4u::Engine& engine,
                                   const std::string& coords_dir)
    : coords_dir_(coords_dir) {
    (void)engine;
    load_directory(coords_dir);
}

// ------------------------------------------------------------------ //
// Load directory
// ------------------------------------------------------------------ //

void MobilityManager::load_directory(const std::string& dir_path) {
    if (!fs::exists(dir_path) || !fs::is_directory(dir_path)) {
        XBT_WARN("Mobility dir '%s' does not exist or is not a directory", dir_path.c_str());
        return;
    }

    int loaded = 0;
    for (const auto& entry : fs::directory_iterator(dir_path)) {
        if (!entry.is_regular_file()) continue;
        if (entry.path().extension() != ".csv") continue;

        try {
            MobilityTrace t(entry.path().string());
            XBT_DEBUG("Loaded trace '%s': %zu waypoints [%.1f – %.1f s]",
                      t.device_name().c_str(), t.size(), t.t_min(), t.t_max());
            traces_.emplace(t.device_name(), std::move(t));
            ++loaded;
        } catch (const std::exception& ex) {
            XBT_WARN("Could not load '%s': %s", entry.path().c_str(), ex.what());
        }
    }
    XBT_INFO("MobilityManager: loaded %d traces from '%s'", loaded, dir_path.c_str());
}

// ------------------------------------------------------------------ //
// Query
// ------------------------------------------------------------------ //

bool MobilityManager::has_trace(const std::string& device_name) const noexcept {
    return traces_.find(device_name) != traces_.end();
}

std::optional<MobilityPosition>
MobilityManager::position_at(const std::string& device_name, double sim_t) const noexcept {
    auto it = traces_.find(device_name);
    if (it == traces_.end()) return std::nullopt;
    return it->second.position_at(sim_t);
}

// ------------------------------------------------------------------ //
// Snapshot recording
// ------------------------------------------------------------------ //

void MobilityManager::record_all(double sim_t) {
    std::lock_guard<std::mutex> lock(mutex_);
    for (const auto& [name, trace] : traces_) {
        snapshots_.push_back({name, trace.position_at(sim_t)});
    }
}

void MobilityManager::record(const std::string& device_name, double sim_t) {
    auto it = traces_.find(device_name);
    if (it == traces_.end()) return;
    std::lock_guard<std::mutex> lock(mutex_);
    snapshots_.push_back({device_name, it->second.position_at(sim_t)});
}

// ------------------------------------------------------------------ //
// Periodic actor
// ------------------------------------------------------------------ //

void MobilityManager::start_periodic_actor(simgrid::s4u::Engine& engine,
                                             double interval_s) {
    if (traces_.empty()) return;

    auto* host = engine.get_all_hosts().front();
    XBT_INFO("Starting MobilityRecorder actor on host '%s' (interval=%.1f s)",
             host->get_cname(), interval_s);

    // Determine latest trace timestamp so the actor auto-terminates
    double t_end = 0.0;
    for (const auto& [name, trace] : traces_)
        t_end = std::max(t_end, trace.t_max());
    t_end += interval_s * 2.0;  // small safety buffer

    // Capture by pointer (manager must outlive the simulation)
    MobilityManager* self = this;
    host->add_actor("mobility_recorder",
        [self, interval_s, t_end]() {
            XBT_INFO("MobilityRecorder started (interval=%.1f s, runs until t=%.1f s)",
                     interval_s, t_end);
            while (simgrid::s4u::Engine::get_clock() <= t_end) {
                double t = simgrid::s4u::Engine::get_clock();
                self->record_all(t);
                simgrid::s4u::this_actor::sleep_for(interval_s);
            }
            XBT_INFO("MobilityRecorder finished");
        });
}

// ------------------------------------------------------------------ //
// Export
// ------------------------------------------------------------------ //

void MobilityManager::export_json(const std::string& filename) const {
    std::ofstream f(filename);
    if (!f) throw std::runtime_error("Cannot open '" + filename + "' for writing");

    f << "[\n";
    bool first = true;
    for (const auto& s : snapshots_) {
        if (!first) f << ",\n";
        first = false;
        const auto& p = s.position;
        f << "  {"
          << "\"device\":"    << quote(s.device_name) << ","
          << "\"timestamp\":" << std::fixed << std::setprecision(6) << p.timestamp << ","
          << "\"latitude\":"  << p.latitude  << ","
          << "\"longitude\":" << p.longitude;
        for (const auto& [key, val] : p.extra)
            f << ",\"" << key << "\":" << val;
        f << "}";
    }
    f << "\n]\n";
    XBT_INFO("Exported %zu snapshots to '%s'", snapshots_.size(), filename.c_str());
}

void MobilityManager::export_csv(const std::string& filename) const {
    std::ofstream f(filename);
    if (!f) throw std::runtime_error("Cannot open '" + filename + "' for writing");

    // Collect all extra key names (in insertion/sorted order across all snapshots)
    std::vector<std::string> extra_keys;
    {
        std::set<std::string> seen;
        for (const auto& s : snapshots_)
            for (const auto& [k, _] : s.position.extra)
                if (seen.insert(k).second) extra_keys.push_back(k);
    }

    // Header
    f << "device,timestamp,latitude,longitude";
    for (const auto& k : extra_keys) f << "," << k;
    f << "\n";

    // Rows
    for (const auto& s : snapshots_) {
        const auto& p = s.position;
        f << s.device_name << ","
          << std::fixed << std::setprecision(6)
          << p.timestamp  << "," << p.latitude  << "," << p.longitude;
        for (const auto& k : extra_keys) {
            auto it = p.extra.find(k);
            f << "," << (it != p.extra.end() ? it->second : 0.0);
        }
        f << "\n";
    }
    XBT_INFO("Exported %zu snapshots to '%s'", snapshots_.size(), filename.c_str());
}

void MobilityManager::export_traces_json(const std::string& filename) const {
    std::ofstream f(filename);
    if (!f) throw std::runtime_error("Cannot open '" + filename + "' for writing");

    f << "{\n";
    bool first_trace = true;
    for (const auto& [name, trace] : traces_) {
        if (!first_trace) f << ",\n";
        first_trace = false;
        f << "  " << quote(name) << ": [\n";
        bool first_wp = true;
        for (const auto& p : trace.waypoints()) {
            if (!first_wp) f << ",\n";
            first_wp = false;
            f << "    {"
              << "\"timestamp\":" << std::fixed << std::setprecision(6) << p.timestamp << ","
              << "\"latitude\":"  << p.latitude  << ","
              << "\"longitude\":" << p.longitude;
            for (const auto& [key, val] : p.extra)
                f << ",\"" << key << "\":" << val;
            f << "}";
        }
        f << "\n  ]";
    }
    f << "\n}\n";
    XBT_INFO("Exported raw traces for %zu devices to '%s'",
             traces_.size(), filename.c_str());
}

} // namespace enigma::mobility

File: src/python/enigma/mobility/mobility_manager.py
"""
mobility_manager.py
-------------------
Manages mobility traces for all devices in a SimGrid simulation.

# Quick-start

## 1. Add ``mobility_dir`` property to your platform XML

    <zone id="root" routing="Full">
      <prop id="mobility_dir" value="platforms/coords/"/>
      ...
    </zone>

## 2. Create the manager *after* loading the platform

    import simgrid
    from enigma.mobility import MobilityManager

    e = simgrid.Engine(sys.argv)
    e.load_platform(platform_xml)

    mob = MobilityManager(e)               # reads property from XML
    # -- or --
    mob = MobilityManager(e, coords_dir="platforms/coords/")

## 3. Query positions inside actors

    pos = mob.position_at("edge_cluster_0_node_1", simgrid.Engine.clock)
    if pos:
        simgrid.this_actor.info(f"lat={pos.latitude:.6f} lon={pos.longitude:.6f}")

## 4. Start the periodic snapshot actor (optional)

    mob.start_periodic_actor(e, interval_s=0.5)

## 5. After e.run(), export

    mob.recorder.export_csv("output_snapshots.csv")
    mob.recorder.export_json("output_snapshots.json")
"""
from __future__ import annotations

import glob
import os
import queue
import threading
from typing import Dict, List, Optional

import simgrid

from .mobility_trace import MobilityTrace, MobilityPosition
from .mobility_recorder import MobilityRecorder


class MobilityManager:
    """
    Central manager: loads traces, interpolates positions, records snapshots.

    Parameters
    ----------
    engine : simgrid.Engine
        Running SimGrid engine (must have loaded a platform).
    coords_dir : str, optional
        Directory with ``<device_name>.csv`` files.  If omitted the manager
        reads the ``mobility_dir`` zone/host property from the platform.
    visualizer : MobilityVisualizer, optional
        If provided, position updates are forwarded there for live display.
    """

    def __init__(
        self,
        engine: simgrid.Engine,
        coords_dir: Optional[str] = None,
        visualizer=None,
        platform_xml: Optional[str] = None,
    ) -> None:
        self._engine     = engine
        self._traces:    Dict[str, MobilityTrace] = {}
        self._recorder   = MobilityRecorder()
        self._visualizer = visualizer
        self._lock       = threading.Lock()

        dir_path = coords_dir or self._detect_mobility_dir(engine, platform_xml)
        if not dir_path:
            print("[MobilityManager] No mobility_dir found – module inactive")
        else:
            self._load_directory(dir_path)

    # ------------------------------------------------------------------ #
    # Properties                                                            #
    # ------------------------------------------------------------------ #

    @property
    def recorder(self) -> MobilityRecorder:
        return self._recorder

    @property
    def trace_count(self) -> int:
        return len(self._traces)

    @property
    def device_names(self) -> List[str]:
        return list(self._traces.keys())

    # ------------------------------------------------------------------ #
    # Queries                                                               #
    # ------------------------------------------------------------------ #

    def has_trace(self, device_name: str) -> bool:
        return device_name in self._traces

    def position_at(self, device_name: str, sim_t: float) -> Optional[MobilityPosition]:
        """
        Return the interpolated position for *device_name* at simulation
        time *sim_t*, or ``None`` if no trace is loaded for that device.
        """
        trace = self._traces.get(device_name)
        if trace is None:
            return None
        return trace.position_at(sim_t)

    def get_trace(self, device_name: str) -> Optional[MobilityTrace]:
        return self._traces.get(device_name)

    # ------------------------------------------------------------------ #
    # Recording helpers (called from within SimGrid actors)                #
    # ------------------------------------------------------------------ #

    def record(self, device_name: str, sim_t: float) -> Optional[MobilityPosition]:
        """
        Compute and record position for one device.
        Returns the position (or None if no trace).
        Thread-safe.
        """
        pos = self.position_at(device_name, sim_t)
        if pos is None:
            return None
        with self._lock:
            self._recorder.add(device_name, pos)
            if self._visualizer is not None:
                self._visualizer.push(device_name, pos)
        return pos

    def record_all(self, sim_t: float) -> Dict[str, MobilityPosition]:
        """
        Compute and record positions for all loaded devices at *sim_t*.
        Returns a dict {device_name: MobilityPosition}.
        Thread-safe.
        """
        result: Dict[str, MobilityPosition] = {}
        with self._lock:
            for name, trace in self._traces.items():
                pos = trace.position_at(sim_t)
                self._recorder.add(name, pos)
                result[name] = pos
                if self._visualizer is not None:
                    self._visualizer.push(name, pos)
        return result

    # ------------------------------------------------------------------ #
    # Periodic actor                                                        #
    # ------------------------------------------------------------------ #

    def start_periodic_actor(
        self,
        engine: simgrid.Engine,
        interval_s: float = 1.0,
    ) -> None:
        """
        Deploy a SimGrid actor that calls ``record_all()`` every
        *interval_s* simulated seconds.  Must be called before ``e.run()``.
        The actor automatically stops once the simulation clock passes the
        latest timestamp in any loaded trace (+ one extra interval).
        """
        if not self._traces:
            return

        # Determine the latest timestamp across all traces
        t_end = max(t.t_max for t in self._traces.values()) + interval_s * 2

        host = engine.all_hosts[0]
        manager = self   # capture reference

        def _recorder_actor():
            simgrid.this_actor.info(
                f"[MobilityRecorder] started  interval={interval_s:.2f} s  "
                f"runs until t={t_end:.1f} s"
            )
            while simgrid.Engine.clock <= t_end:
                t = simgrid.Engine.clock
                manager.record_all(t)
                simgrid.this_actor.sleep_for(interval_s)
            simgrid.this_actor.info("[MobilityRecorder] finished")

        simgrid.Actor.create("mobility_recorder", host, _recorder_actor)

    # ------------------------------------------------------------------ #
    # Internal                                                              #
    # ------------------------------------------------------------------ #

    @staticmethod
    def _detect_mobility_dir(engine: simgrid.Engine,
                              platform_xml: Optional[str] = None) -> str:
        """Return the value of the 'mobility_dir' property from any zone/host.

        SimGrid's Python bindings expose ``host.get_property()`` but **zone**
        properties are not propagated down to hosts.  We therefore parse the
        XML file directly when the path is known.
        """
        # 1. Try host-level properties (works if prop is on a host, not a zone)
        for host in engine.all_hosts:
            try:
                val = host.get_property("mobility_dir")
                if val:
                    return val
            except Exception:
                pass

        # 2. Parse the XML directly – handles zone-level <prop> elements
        if platform_xml and os.path.isfile(platform_xml):
            try:
                import xml.etree.ElementTree as ET
                tree = ET.parse(platform_xml)
                for elem in tree.iter():
                    if elem.tag == "prop" and elem.get("id") == "mobility_dir":
                        return elem.get("value", "")
            except Exception:
                pass

        return ""

    def _load_directory(self, dir_path: str) -> None:
        if not os.path.isdir(dir_path):
            print(f"[MobilityManager] Directory not found: {dir_path}")
            return

        csv_files = glob.glob(os.path.join(dir_path, "*.csv"))
        if not csv_files:
            print(f"[MobilityManager] No CSV files found in {dir_path}")
            return

        loaded = 0
        for csv_path in sorted(csv_files):
            try:
                t = MobilityTrace(csv_path)
                self._traces[t.device_name] = t
                loaded += 1
            except Exception as ex:
                print(f"[MobilityManager] Warning: could not load {csv_path}: {ex}")

        print(
            f"[MobilityManager] Loaded {loaded} traces from '{dir_path}'"
        )
